Crashes in Command Mode

Hi there,

first of all thank you very much for the Quarkus JBeret Extension. 

I ran into an issue. Since I haven't worked much with neither Quarkus nor JBeret I am unsure if this is really an issue at all or just my misunderstand of things.

My pet project breaks if I treat it like a command mode application. Motivation: run a job once and the forget about it.

This is my code:

```
@QuarkusMain
public class Main implements QuarkusApplication {

    @Inject
    QuarkusJobOperator jobOperator;

    public static void main(String... args) {
        Quarkus.run(Main.class, args);
    }

    @Override
    public int run(String[] args) throws Exception {
        long executionId = jobOperator.start("numbers", new Properties());
        return 0;
    }

}
```

If I run the Java main method I get (among others) the following error message:

```
2026-05-19 15:03:15,802 ERROR [org.jberet] (executor-thread-4) JBERET000007: Failed to run job numbers, sum, org.jberet.job.model.Chunk@26c4b2b7 [Error Occurred After Shutdown]: java.lang.IllegalStateException: JBERET000600: Failed to create artifact with ref name MyReader.  Ensure CDI beans.xml is present and batch.xml, if any, is configured properly.
        at org.jberet.runtime.context.JobContextImpl.createArtifact(JobContextImpl.java:197)
        at org.jberet.runtime.runner.AbstractRunner.createArtifact(AbstractRunner.java:154)
        at org.jberet.runtime.runner.ChunkRunner.run(ChunkRunner.java:139)
        at org.jberet.runtime.runner.StepExecutionRunner.runBatchletOrChunk(StepExecutionRunner.java:223)
        at org.jberet.runtime.runner.StepExecutionRunner.run(StepExecutionRunner.java:142)
        at org.jberet.runtime.runner.CompositeExecutionRunner.runStep(CompositeExecutionRunner.java:170)
        at org.jberet.runtime.runner.CompositeExecutionRunner.runJobElement(CompositeExecutionRunner.java:134)
        at org.jberet.runtime.runner.StepExecutionRunner.run(StepExecutionRunner.java:200)
        at org.jberet.runtime.runner.CompositeExecutionRunner.runStep(CompositeExecutionRunner.java:170)
        at org.jberet.runtime.runner.CompositeExecutionRunner.runFromHeadOrRestartPoint(CompositeExecutionRunner.java:94)
        at org.jberet.runtime.runner.JobExecutionRunner.run(JobExecutionRunner.java:58)
        at org.jberet.spi.JobExecutor$1.run(JobExecutor.java:100)
        at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:682)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1586)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkus.arc.ArcContainer.beanManager()" because the return value of "io.quarkus.arc.Arc.container()" is null
```

However, if I start the job from within Application Scoped Bean it runs without errors:

```
@ApplicationScoped
public class Scheduler {

    @Inject
    QuarkusJobOperator jobOperator;

    void onStart(@Observes StartupEvent startupEvent) {
        jobOperator.start("numbers", new Properties());
    }
}
```

Not sure if this is intended.

For reference, here is my numbers.xml file and the MyReader Java class. I tried different annotations with MyReader: Dependent, Application oder Singleton. All show the same error message.

```
<?xml version="1.0" encoding="UTF-8"?>
<job id="numbers" xmlns="https://jakarta.ee/xml/ns/jakartaee"
     version="2.0">

    <step id="start" next="sum">
        <batchlet ref="Snapshot"></batchlet>
    </step>
    <step id="sum">
        <chunk item-count="1">
            <skippable-exception-classes>
                <include class="java.lang.Exception"/>
            </skippable-exception-classes>
            <reader ref="MyReader"></reader>
            <processor ref="MyProcessor"></processor>
            <writer ref="MyWriter"></writer>
        </chunk>
    </step>
</job>
```

MyReader.java
```
@Named
public class MyReader implements ItemReader {

    private MyCheckpoint checkpoint;

    @Inject
    JobContext context;

    /* count of numbers */
    int n = 100;
    /* array of numbers to work with */
    int[] numbers;
    /* index to iterate through numbers */
    int i;

    @Override
    public void open(Serializable ckpt) throws Exception {
        if (ckpt == null) {
            checkpoint = new MyCheckpoint();
        } else {
            checkpoint = (MyCheckpoint) ckpt;
        }

        numbers = new int[n];
        for (int i = 0; i < numbers.length; i++) {
            numbers[i] = i;
        }

        i = 0;
    }

    @Override
    public void close() {}

    @Override
    public Object readItem() {
        if (i < numbers.length) {
            return numbers[i++];
        } else {
            return null;
        }
    }

    @Override
    public Serializable checkpointInfo()  {
        return checkpoint;
    }
}
```

--- a/core/deployment/src/main/java/io/quarkiverse/jberet/deployment/JBeretProcessor.java
+++ b/core/deployment/src/main/java/io/quarkiverse/jberet/deployment/JBeretProcessor.java
@@ -57,6 +57,7 @@ import io.quarkiverse.jberet.runtime.JBeretRecorder;
 import io.quarkiverse.jberet.runtime.JobProcessor;
 import io.quarkiverse.jberet.runtime.JobProcessor.JobProcessorBuilder;
 import io.quarkiverse.jberet.runtime.JobsProducer;
+import io.quarkiverse.jberet.runtime.QuarkusJobLauncher;
 import io.quarkiverse.jberet.runtime.QuarkusJobRepository;
 import io.quarkiverse.jberet.runtime.QuarkusJobScheduler;
 import io.quarkiverse.jberet.runtime.repository.InMemoryJobRepositorySupplier;
@@ -136,6 +137,7 @@ class JBeretProcessor {
         additionalBeans.produce(new AdditionalBeanBuildItem(JBeretProducer.class));
         additionalBeans.produce(new AdditionalBeanBuildItem(JobsProducer.class));
         additionalBeans.produce(new AdditionalBeanBuildItem(QuarkusJobScheduler.class));
+        additionalBeans.produce(new AdditionalBeanBuildItem(QuarkusJobLauncher.class));
 
         additionalBeans.produce(new AdditionalBeanBuildItem(QuarkusJobRepository.class));
         additionalBeans.produce(new AdditionalBeanBuildItem(InMemoryJobRepositorySupplier.class));

--- a/core/runtime/src/main/java/io/quarkiverse/jberet/runtime/QuarkusJobLauncher.java
+++ b/core/runtime/src/main/java/io/quarkiverse/jberet/runtime/QuarkusJobLauncher.java
@@ -1,0 +1,58 @@
+package io.quarkiverse.jberet.runtime;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+import jakarta.batch.operations.JobOperator;
+import jakarta.batch.runtime.BatchStatus;
+import jakarta.batch.runtime.JobExecution;
+import jakarta.batch.runtime.JobInstance;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+import org.jberet.runtime.JobExecutionImpl;
+
+@ApplicationScoped
+public class QuarkusJobLauncher {
+    @Inject
+    JobOperator jobOperator;
+
+    /**
+     * Starts or restarts a job synchronously. If the most recent execution of the job is
+     * {@link BatchStatus#FAILED} or {@link BatchStatus#STOPPED}, the job is restarted from its
+     * last checkpoint. Otherwise, a new job instance is created. The method blocks until the job
+     * completes.
+     *
+     * @param jobName the name of the job XML definition
+     * @param jobParameters the job parameters
+     * @return the completed {@link JobExecution}
+     */
+    public JobExecution run(String jobName, Properties jobParameters) throws Exception {
+        long executionId = startOrRestart(jobName, jobParameters);
+
+        JobExecutionImpl jobExecution = (JobExecutionImpl) jobOperator.getJobExecution(executionId);
+        jobExecution.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+
+        return jobExecution;
+    }
+
+    private long startOrRestart(String jobName, Properties jobParameters) {
+        List<JobInstance> instances = jobOperator.getJobInstances(jobName, 0, 1);
+        if (!instances.isEmpty()) {
+            List<JobExecution> executions = jobOperator.getJobExecutions(instances.get(0));
+            if (!executions.isEmpty()) {
+                JobExecution lastExecution = executions.stream()
+                        .max(Comparator.comparingLong(JobExecution::getExecutionId))
+                        .get();
+                BatchStatus status = lastExecution.getBatchStatus();
+                if (status == BatchStatus.FAILED || status == BatchStatus.STOPPED) {
+                    return jobOperator.restart(lastExecution.getExecutionId(), jobParameters);
+                }
+            }
+        }
+
+        return jobOperator.start(jobName, jobParameters);
+    }
+}

--- a/docs/modules/ROOT/pages/index.adoc
+++ b/docs/modules/ROOT/pages/index.adoc
@@ -153,6 +153,54 @@ public static class JobProducer {
 A `Job` registered with CDI will be named by the `@Named#value` attribute or by the method name. The
 `@Named` annotation is required regardless. This is the named used by `JobOperator` to locate and execute a `Job`.
 
+== Job Launcher
+
+The `QuarkusJobLauncher` provides a simplified way to start and run batch jobs synchronously. Unlike the standard
+`JobOperator`, which always starts jobs asynchronously, the `QuarkusJobLauncher` blocks until the job completes and
+returns the final `JobExecution`:
+
+[source,java]
+----
+@Inject
+QuarkusJobLauncher jobLauncher;
+
+void execute() throws Exception {
+    JobExecution execution = jobLauncher.run("myJob", new Properties());
+    // Execution is already complete here
+    BatchStatus status = execution.getBatchStatus();
+}
+----
+
+The `run` method also handles restart automatically. If the most recent execution of the job is `FAILED` or `STOPPED`,
+calling `run` will restart it from the last checkpoint instead of creating a new job instance. If the job has never been
+run or the last execution completed successfully, a new job instance is created.
+
+=== Command Mode with `@QuarkusMain`
+
+The `QuarkusJobLauncher` is designed to work with Quarkus
+https://quarkus.io/guides/lifecycle#the-main-method[Command Mode]. This allows you to run a batch job as a CLI
+application that starts, executes the job, and exits:
+
+[source,java]
+----
+@QuarkusMain
+public class BatchApplication implements QuarkusApplication {
+    @Inject
+    QuarkusJobLauncher jobLauncher;
+
+    @Override
+    public int run(String... args) throws Exception {
+        Properties jobParameters = new Properties();
+        // args can be used to set job parameters
+        JobExecution execution = jobLauncher.run("myJob", jobParameters);
+        return execution.getBatchStatus() == BatchStatus.COMPLETED ? 0 : 1;
+    }
+}
+----
+
+The application will start Quarkus, execute the batch job, and return an exit code based on the job result. This is
+useful for scheduled batch processing via cron, CI/CD pipelines, or any scenario where the job should run once and exit.
+
 // TODO - Rewrite
 == Scheduler
 
@@ -252,6 +300,7 @@ Example applications can be found inside the `integration-tests` folder:
 - `chunk` - A simple Job that reads, processes, and stores data from a file.
 - `jdbc-repository` - A Job that uses a `jdbc` datasource to store JBeret and Job metadata.
 - `scheduler` - Schedule a Job to run every 10 seconds
+- `job-launcher` - A `@QuarkusMain` application using `QuarkusJobLauncher` with start and restart
 
 Or take a look into the https://github.com/radcortez/wow-auctions[World of Warcraft Auctions - Batch Application]. It
 downloads the World of Warcraft Auction House data and provides statistics about items prices.

--- a/integration-tests/job-launcher/pom.xml
+++ b/integration-tests/job-launcher/pom.xml
@@ -1,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>io.quarkiverse.jberet</groupId>
+    <artifactId>quarkus-jberet-integration-tests</artifactId>
+    <version>2.10.1-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>quarkus-jberet-integration-tests-job-launcher</artifactId>
+  <name>Quarkus - JBeret - Integrations Tests - Job Launcher</name>
+
+</project>

--- a/integration-tests/job-launcher/src/main/java/io/quarkiverse/jberet/it/joblauncher/BatchApplication.java
+++ b/integration-tests/job-launcher/src/main/java/io/quarkiverse/jberet/it/joblauncher/BatchApplication.java
@@ -1,0 +1,40 @@
+package io.quarkiverse.jberet.it.joblauncher;
+
+import java.util.Properties;
+
+import jakarta.batch.runtime.BatchStatus;
+import jakarta.batch.runtime.JobExecution;
+import jakarta.inject.Inject;
+
+import io.quarkiverse.jberet.runtime.QuarkusJobLauncher;
+import io.quarkus.runtime.QuarkusApplication;
+import io.quarkus.runtime.annotations.QuarkusMain;
+
+@QuarkusMain
+public class BatchApplication implements QuarkusApplication {
+    @Inject
+    QuarkusJobLauncher jobLauncher;
+
+    @Override
+    public int run(String... args) throws Exception {
+        if (args.length > 0 && "restart".equals(args[0])) {
+            return restart();
+        }
+        return start();
+    }
+
+    private int start() throws Exception {
+        JobExecution execution = jobLauncher.run("main", new Properties());
+        return execution.getBatchStatus() == BatchStatus.COMPLETED ? 0 : 1;
+    }
+
+    private int restart() throws Exception {
+        JobExecution firstExecution = jobLauncher.run("restart", new Properties());
+        if (firstExecution.getBatchStatus() != BatchStatus.FAILED) {
+            return 1;
+        }
+
+        JobExecution restartExecution = jobLauncher.run("restart", new Properties());
+        return restartExecution.getBatchStatus() == BatchStatus.COMPLETED ? 0 : 1;
+    }
+}

--- a/integration-tests/job-launcher/src/main/java/io/quarkiverse/jberet/it/joblauncher/MainBatchlet.java
+++ b/integration-tests/job-launcher/src/main/java/io/quarkiverse/jberet/it/joblauncher/MainBatchlet.java
@@ -1,0 +1,15 @@
+package io.quarkiverse.jberet.it.joblauncher;
+
+import jakarta.batch.api.AbstractBatchlet;
+import jakarta.batch.runtime.BatchStatus;
+import jakarta.enterprise.context.Dependent;
+import jakarta.inject.Named;
+
+@Named
+@Dependent
+public class MainBatchlet extends AbstractBatchlet {
+    @Override
+    public String process() {
+        return BatchStatus.COMPLETED.toString();
+    }
+}

--- a/integration-tests/job-launcher/src/main/java/io/quarkiverse/jberet/it/joblauncher/RestartBatchlet.java
+++ b/integration-tests/job-launcher/src/main/java/io/quarkiverse/jberet/it/joblauncher/RestartBatchlet.java
@@ -1,0 +1,22 @@
+package io.quarkiverse.jberet.it.joblauncher;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import jakarta.batch.api.AbstractBatchlet;
+import jakarta.batch.runtime.BatchStatus;
+import jakarta.enterprise.context.Dependent;
+import jakarta.inject.Named;
+
+@Named
+@Dependent
+public class RestartBatchlet extends AbstractBatchlet {
+    static final AtomicBoolean FIRST_RUN = new AtomicBoolean(true);
+
+    @Override
+    public String process() {
+        if (FIRST_RUN.compareAndSet(true, false)) {
+            throw new RuntimeException("Simulated failure on first run");
+        }
+        return BatchStatus.COMPLETED.toString();
+    }
+}

--- a/integration-tests/job-launcher/src/main/resources/META-INF/batch-jobs/main.xml
+++ b/integration-tests/job-launcher/src/main/resources/META-INF/batch-jobs/main.xml
@@ -1,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<job id="main" xmlns="http://xmlns.jcp.org/xml/ns/javaee" version="1.0">
+  <step id="mainStep">
+    <batchlet ref="mainBatchlet"/>
+  </step>
+</job>

--- a/integration-tests/job-launcher/src/main/resources/META-INF/batch-jobs/restart.xml
+++ b/integration-tests/job-launcher/src/main/resources/META-INF/batch-jobs/restart.xml
@@ -1,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<job id="restart" xmlns="http://xmlns.jcp.org/xml/ns/javaee" version="1.0" restartable="true">
+  <step id="restartStep">
+    <batchlet ref="restartBatchlet"/>
+  </step>
+</job>

--- a/integration-tests/job-launcher/src/test/java/io/quarkiverse/jberet/it/joblauncher/BatchApplicationIT.java
+++ b/integration-tests/job-launcher/src/test/java/io/quarkiverse/jberet/it/joblauncher/BatchApplicationIT.java
@@ -1,0 +1,7 @@
+package io.quarkiverse.jberet.it.joblauncher;
+
+import io.quarkus.test.junit.main.QuarkusMainIntegrationTest;
+
+@QuarkusMainIntegrationTest
+public class BatchApplicationIT extends BatchApplicationTest {
+}

--- a/integration-tests/job-launcher/src/test/java/io/quarkiverse/jberet/it/joblauncher/BatchApplicationTest.java
+++ b/integration-tests/job-launcher/src/test/java/io/quarkiverse/jberet/it/joblauncher/BatchApplicationTest.java
@@ -1,0 +1,24 @@
+package io.quarkiverse.jberet.it.joblauncher;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.main.Launch;
+import io.quarkus.test.junit.main.LaunchResult;
+import io.quarkus.test.junit.main.QuarkusMainTest;
+
+@QuarkusMainTest
+class BatchApplicationTest {
+    @Test
+    @Launch
+    void main(LaunchResult result) {
+        assertEquals(0, result.exitCode());
+    }
+
+    @Test
+    @Launch("restart")
+    void restart(LaunchResult result) {
+        assertEquals(0, result.exitCode());
+    }
+}

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -71,6 +71,7 @@
     <module>scripting</module>
     <module>scopes</module>
     <module>mailer</module>
+    <module>job-launcher</module>
     <module>mongo</module>
     <module>tck</module>
   </modules>

